### PR TITLE
Add integration points for jib auto sync.

### PR DESCRIPTION
**Pull request by loosebazooka**
Code for
- initializing sync state
- calculating sync differences
- gradle/maven determination

The meat of this PR is the logic in `GetSyncDiff`

### CODE REVIEW

File: pkg/skaffold/build/jib/sync.go [hunk 1]
```diff
@@ -89,3 +221,14 @@ func (sm SyncMap) addEntries(entries []JSONSyncEntry, direct bool) error {
 	}
 	return nil
 }
+
+func toAbs(f string) (string, error) {
+	if !filepath.IsAbs(f) {
```
Comment (tejal29) [inline pkg/skaffold/build/jib/sync.go]: Can we reuse this util function ?https://github.com/GoogleContainerTools/skaffold/blob/2ba2af8060e4f32684f9f53989039b7f7588aeca/pkg/skaffold/util/util.go#L238

File: pkg/skaffold/build/jib/sync.go [hunk 2]
```diff
@@ -46,6 +53,131 @@ type JSONSyncEntry struct {
...
+}
+
+// returns toCopy, toDelete, error
+func GetSyncDiff(ctx context.Context, workspace string, a *latest.JibArtifact, e filemon.Events) (map[string][]string, map[string][]string, error) {
+	// if anything that was modified was a buildfile, do NOT sync, do a rebuild
+	buildFiles := GetBuildDefinitions(workspace, a)
+	for _, f := range e.Modified {
+		f, err := toAbs(f)
+		if err != nil {
+			return nil, nil, errors.WithStack(err)
```
Comment (tejal29) [inline pkg/skaffold/build/jib/sync.go]: We have not used `errors.WithStack` elsewhere in the code. Can we use `errors.Wrap` like elsewhere here?

File: pkg/skaffold/build/jib/sync.go [hunk 3]
```diff
@@ -46,6 +53,131 @@ type JSONSyncEntry struct {
...
+		}
+		for _, bf := range buildFiles {
+			if f == bf {
+				return nil, nil, nil
+			}
+		}
+	}
+
+	// no deletions
+	if len(e.Deleted) != 0 {
```
Comment (tejal29) [inline pkg/skaffold/build/jib/sync.go]: Move this code up and exit faster?

File: pkg/skaffold/build/jib/sync.go (hunk 1 above)
Comment (loosebazooka) [reply pkg/skaffold/build/jib/sync.go]: In jib, files can be outside of the workspace potentially.

File: pkg/skaffold/build/jib/sync.go [hunk 4]
```diff
@@ -46,6 +53,131 @@ type JSONSyncEntry struct {
...
+		// change into logging
+		fmt.Println("Deletions are not supported by jib auto sync at the moment")
+		return nil, nil, nil
+	}
+
+	currSyncMap := syncLists[getProjectKey(workspace, a)]
+
+	// if all files are modified and direct, we don't need to build anything
```
Comment (tejal29) [inline pkg/skaffold/build/jib/sync.go]: i did not understand the comment. did you mean "all files modified are direct"?

File: pkg/skaffold/build/jib/sync.go [hunk 5]
```diff
@@ -46,6 +53,131 @@ type JSONSyncEntry struct {
...
+	}
+
+	// we need to do another build and get a new sync map
+	nextSyncMap, err := getSyncMapFunc(ctx, workspace, a)
+	if err != nil {
+		return nil, nil, err
+	}
+	syncLists[getProjectKey(workspace, a)] = *nextSyncMap
+
+	fmt.Println("curr", currSyncMap)
```
Comment (tejal29) [inline pkg/skaffold/build/jib/sync.go]: remove `fmt.Println`

File: pkg/skaffold/build/jib/sync.go [hunk 6]
```diff
@@ -46,6 +53,131 @@ type JSONSyncEntry struct {
...
+		return nil, err
+	}
+
+	switch t {
+	case JibMaven:
+		return getSyncMapCommandMaven(ctx, workspace, artifact), nil
+	case JibGradle:
+		return getSyncMapCommandGradle(ctx, workspace, artifact), nil
+	default:
+		return nil, errors.Errorf("unable to determine Jib builder type for %s", workspace)
```
Comment (tejal29) [inline pkg/skaffold/build/jib/sync.go]: Can you also add the plugin type "t" ?

File: pkg/skaffold/build/jib/sync.go (hunk 4 above)
Comment (loosebazooka) [reply pkg/skaffold/build/jib/sync.go]: updated the comment

File: pkg/skaffold/build/jib/sync.go [hunk 7]
```diff
@@ -46,6 +53,132 @@ type JSONSyncEntry struct {
...
+	syncLists[getProjectKey(workspace, a)] = *syncMap
+	return nil
+}
+
+// returns toCopy, toDelete, error
+func GetSyncDiff(ctx context.Context, workspace string, a *latest.JibArtifact, e filemon.Events) (map[string][]string, map[string][]string, error) {
+	// no deletions allowed
+	if len(e.Deleted) != 0 {
+		// change into logging
+		fmt.Println("Deletions are not supported by jib auto sync at the moment")
```
Comment (loosebazooka) [inline pkg/skaffold/build/jib/sync.go]: I still need to fix this to use logger

**Pull request merged**